.github/workflows/ci.yaml

```diff
@@ -1,20 +1,536 @@
+      run: |
+        user="test"
+        pass="test"
+
+        echo "Generating users"
+        htpasswd -Bbn "$user" "$pass" >> htpasswd # authenticated user for tests
+
+        # the registry can be accessed from inside the kind cluster via it's name.
+        # however, due to a quirk with ggcr, it will force https connections UNLESS
+        # it fits certain patterns (https://github.com/google/go-containerregistry/blob/b8d1c0a1df12a3b3dd3e7f98330780fdc015ce40/pkg/name/registry.go#L75)
+        # so in order to avoid setting tls for the registry, we can name it something.local
+        # to trick ggcr into using the http scheme
+        name="registry.local"
+
+        echo "Starting registry"
+        if [ "$(docker inspect -f '{{.State.Running}}' "${reg_name}" 2>/dev/null || true)" != 'true' ]; then
+          docker run --detach \
+            --name "$name" \
+            -v "$(pwd)/htpasswd:/auth/htpasswd" \
+            -e "REGISTRY_AUTH=htpasswd" \
+            -e "REGISTRY_AUTH_HTPASSWD_REALM=Registry Realm" \
+            -e "REGISTRY_AUTH_HTPASSWD_PATH=/auth/htpasswd" \
+            -e "REGISTRY_STORAGE_DELETE_ENABLED=true" \
+            --publish "5000:5000" \
+            registry:2
+        fi
+
+        echo "REGISTRY_URL=$name:5000" >> $GITHUB_ENV
+        echo "REGISTRY_USER=$user" >> $GITHUB_ENV
+        echo "REGISTRY_PASS=$pass" >> $GITHUB_ENV
+
+        # alias registry.local to localhost
+        # ip=$(docker container inspect $name --format '{{ .NetworkSettings.Networks.kind.IPAddress }}')
+        ip=127.0.0.1
+        echo "$ip registry.local" | sudo tee -a /etc/hosts
+
+        cat <<EOF > kind.yaml
+        kind: Cluster
+        apiVersion: kind.x-k8s.io/v1alpha4
+        containerdConfigPatches:
+        - |-
+          [plugins."io.containerd.grpc.v1.cri".registry.mirrors."registry.local:5000"]
+            endpoint = ["http://registry.local:5000"]
+        EOF
+
+    - name: Create Kind Cluster
+      uses: helm/kind-action@v1.8.0
+      with:
+        cluster_name: e2e
+        config: kind.yaml
+
+    - name: Docker Login
+      uses: docker/login-action@v2.2.0
+      with:
+        registry: ${{ env.REGISTRY_URL }}
+        username: ${{ env.REGISTRY_USER }}
+        password: ${{ env.REGISTRY_PASS }}
+
+    - name: Run tests
+      run: |
+        # make the registry container accessible by name from inside the cluster
+        docker network connect kind registry.local
+
+        kapp deploy -a kpack -y -f prerelease.yaml
+
+        export IMAGE_REGISTRY=${{ env.REGISTRY_URL }}
+        export IMAGE_REGISTRY_USERNAME=${{ env.REGISTRY_USER }}
+        export IMAGE_REGISTRY_PASSWORD=${{ env.REGISTRY_PASS }}
+
+        export GIT_PRIVATE_REPO=${{ vars.E2E_PRIVATE_REPO }}
+        export GIT_BASIC_USERNAME="${{ secrets.E2E_PRIVATE_REPO_USERNAME }}"
+        export GIT_BASIC_PASSWORD="${{ secrets.E2E_PRIVATE_REPO_PASSWORD }}"
+        export GIT_SSH_PRIVATE_KEY="${{ secrets.E2E_PRIVATE_REPO_PRIVATE_KEY }}"
```

Comment:
These aren't needed in <0.11.x, but there's no real harm in having them here I guess

Reply:
I can remove them